### PR TITLE
Pipeline: Add JSON Schema Support

### PR DESCRIPTION
**Pull request by georg-schwarz**

### CODE REVIEW

File: pipeline/src/pipeline-config/model/pipelineConfig.ts [hunk 1]
```diff
@@ -5,7 +5,7 @@ export interface PipelineConfig {
   datasourceId: number
   transformation: TransformationConfig
   metadata: Metadata
-  schema?: object
+  schema: object
```
Comment (georg-schwarz) [inline pipeline/src/pipeline-config/model/pipelineConfig.ts]: Schould be left optional (same as in adapter) => we will only provide a good query API if schema is given.

File: pipeline/src/pipeline-config/model/pipelineTransformedData.ts [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+export interface PipelineTransformedData {
+  id: number
+  pipelineId: number
+  healthStatus: string
+  data: unknown
+  schema: object
+}
+
+export interface PipelineTransformedDataDTO {
```
Comment (georg-schwarz) [inline pipeline/src/pipeline-config/model/pipelineTransformedData.ts]: You can use a Utility Type: `type PipelineTransformedDataDTO = Omit<PipelineTransformedData, "id">`

File: pipeline/src/pipeline-config/pipelineConfigManager.test.ts [hunk 3]
```diff
@@ -7,6 +7,7 @@ import PipelineExecutor from '../pipeline-execution/pipelineExecutor'
 import { PipelineConfig, PipelineConfigDTO } from './model/pipelineConfig'
 import VM2SandboxExecutor from '../pipeline-execution/sandbox/vm2SandboxExecutor'
 import * as PipelineConfigRepository from './pipelineConfigRepository'
+// import * as PipelineTransformedDataRepository from './pipelineTransformedDataRepository'
```
Comment (georg-schwarz) [inline pipeline/src/pipeline-config/pipelineConfigManager.test.ts]: remove if not used

File: pipeline/src/pipeline-config/pipelineConfigManager.ts [hunk 4]
```diff
@@ -59,12 +62,33 @@ export class PipelineConfigManager {
...
+    const ajv = new Ajv({ strict: false })
     for (const config of allConfigs) {
       const result = this.pipelineExecutor.executeJob(config.transformation.func, data)
       if ('error' in result) {
         await this.pgClient.transaction(async client =>
           await EventPublisher.publishError(client, config.id, config.metadata.displayName, result.error.message))
       } else if ('data' in result) {
+        let validate: any
+        let valid: boolean = true
+        if (config.schema !== undefined || config.schema !== null) {
```
Comment (georg-schwarz) [inline pipeline/src/pipeline-config/pipelineConfigManager.ts]: outsource the whole validation in own method

**Review by georg-schwarz — COMMENTED** [review]
Left some minor comments :)

File: docker-compose.yml [hunk 5]
```diff
@@ -171,10 +171,10 @@ services:
   #   - "./pgdata:/var/lib/postgresql/data"
 
   # Uncomment this if you want to manage the Postgres databases with adminer
-  # adminer: # management UI for Postgres
-  #   image: adminer
-  #   ports:
-  #     - 8080:8080
+  adminer: # management UI for Postgres
```
Comment (georg-schwarz) [inline docker-compose.yml]: Pls comment out again

File: notification/src/api/pipelineEvent.ts [hunk 6]
```diff
@@ -4,6 +4,7 @@ export interface PipelineSuccessEvent {
   pipelineId: number
   pipelineName: string
   data: object
+  schema: object
```
Comment (georg-schwarz) [inline notification/src/api/pipelineEvent.ts]: Should be optional

File: pipeline/src/pipeline-config/model/pipelineTransformedData.ts [hunk 7]
```diff
@@ -0,0 +1,61 @@
+import { validators } from '@jvalue/node-dry-basics'
+
+export interface PipelineTransformedData {
+  id: number
+  pipelineId: number
+  healthStatus: string
```
Comment (georg-schwarz) [inline pipeline/src/pipeline-config/model/pipelineTransformedData.ts]: Should be an enum

File: pipeline/src/pipeline-config/outboxEventPublisher.ts [hunk 8]
```diff
@@ -41,11 +41,18 @@ export async function publishError
...
     pipelineName: pipelineName,
     data: result
   }
+  if (schema !== undefined || schema !== null) {
+    content = {
+      ...content,
+      schema: schema as object
+    }
+  }
+  console.log('**********PIPELOINEEXECUTIONSUCCESS************')
```
Comment (georg-schwarz) [inline pipeline/src/pipeline-config/outboxEventPublisher.ts]: remove debugging prints

File: pipeline/src/pipeline-config/outboxEventPublisher.ts [hunk 9]
```diff
@@ -41,11 +41,18 @@ export async function publishError
...
+(client: ClientBase, pipelineId: number, pipelineName: string, result: unknown, schema?: object): Promise<string> {
+  let content: any = {
     pipelineId: pipelineId,
     pipelineName: pipelineName,
     data: result
   }
+  if (schema !== undefined || schema !== null) {
+    content = {
+      ...content,
+      schema: schema as object
```
Comment (georg-schwarz) [inline pipeline/src/pipeline-config/outboxEventPublisher.ts]: You could make it shorter by
```
const content = {
    pipelineId: pipelineId,
    pipelineName: pipelineName,
    data: result,
    schema: schema != null ? schema as object : undefined
}
```

or just do
```
content.schema = schema as obejct;
``` 
and provide a schema for `content` (pls don't use any)

File: storage/storage-mq/src/api/pipelineExecutionEventHandler.ts [hunk 10]
```diff
@@ -16,5 +16,6 @@ export interface PipelineExecutedEvent {
   pipelineId: number
   pipelineName: string
   data: unknown
+  schema: object
```
Comment (georg-schwarz) [inline storage/storage-mq/src/api/pipelineExecutionEventHandler.ts]: optional?

File: ui/src/pipeline/pipeline.ts [hunk 11]
```diff
@@ -3,6 +3,7 @@ export default interface Pipeline {
   datasourceId: number
   metadata: PipelineMetaData
   transformation: TransformationConfig
+  schema: object
```
Comment (georg-schwarz) [inline ui/src/pipeline/pipeline.ts]: optional?

**Comment by lunedis** [comment]
Please rebase this of main so we can see updated CI results.

File: pipeline/src/pipeline-validator/jsonschemavalidator.ts [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+    let valid: boolean = true
+
+    const transformedData: PipelineTransformedDataDTO = {
+      pipelineId: config.id,
+      healthStatus: HealthStatus.OK,
+      data: data as object
+    }
+
+    const ajv = new Ajv({ strict: false })
+    if (config.schema !== undefined || config.schema !== null) {
```
Comment (georg-schwarz) [inline pipeline/src/pipeline-validator/jsonschemavalidator.ts]: Currently causes the CI to fail: has to be an `&&`

File: pipeline/src/pipeline-validator/jsonschemavalidator.ts [hunk 13]
```diff
@@ -0,0 +1,33 @@
+import Ajv from 'ajv'
```
Comment (georg-schwarz) [inline pipeline/src/pipeline-validator/jsonschemavalidator.ts]: pls rename file to `JsonSchemaValidator.ts`

Comment (LechoDecho) [reply pipeline/src/pipeline-validator/jsonschemavalidator.ts]: x

**Comment by lunedis** [comment]
I renamed the file, maybe there was a Windows issue? https://github.com/LechoDecho/open-data-service/pull/3

**Pull request merged**